core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -251,4 +322,126 @@ private static void validatePlanningArguments(long splitSize, int lookback, long
+    private final CloseableIterable<T> files;
+    private final int parallelism;
+    private final long splitSize;
+    private final int lookback;
+    private final Function<T, Long> weightFunc;
+    private final BiFunction<CloseableIterable<T>, Long, CloseableIterable<T>> splitFunc;
+    private final Function<List<T>, G> groupFunc;
+
+    private Long targetSize = null;
+
+    private AdaptiveSplitPlanningIterable(
+        CloseableIterable<T> files,
+        int parallelism,
+        long splitSize,
+        int lookback,
+        Function<T, Long> weightFunc,
+        BiFunction<CloseableIterable<T>, Long, CloseableIterable<T>> splitFunc,
+        Function<List<T>, G> groupFunc) {
+      this.files = files;
+      this.parallelism = parallelism;
+      this.splitSize = splitSize;
+      this.lookback = lookback;
+      this.weightFunc = weightFunc;
+      this.splitFunc = splitFunc;
+      this.groupFunc = groupFunc;
+    }
+
+    @Override
+    public CloseableIterator<G> iterator() {
+      if (targetSize != null) {
+        // target size is already known so plan with the static target size
+        CloseableIterable<T> splitTasks = splitFunc.apply(files, targetSize);
+        CloseableIterator<G> iter =
+            planTasksInternal(splitTasks, targetSize, lookback, weightFunc, groupFunc).iterator();
+        addCloseable(iter);
+        return iter;
+      }
+
+      boolean shouldClose = true;
+      CloseableIterator<T> tasksIter = files.iterator();
+      try {
+        // load tasks until the iterator is exhausted or until the total weight is enough to get the
+        // parallelism at the split size passed in.
+        Deque<T> readAheadTasks = Lists.newLinkedList();
+        long readToSize = parallelism * splitSize;
+        long totalSize = 0L;
+
+        while (tasksIter.hasNext()) {
+          T task = tasksIter.next();
+          readAheadTasks.addLast(task);
+          totalSize += weightFunc.apply(task);
+
+          if (totalSize > readToSize) {
+            break;
+          }
+        }
+
+        // if total size was reached, then the requested split size is used. otherwise, the iterator
+        // was exhausted and the split size will be adjusted to target parallelism with a reasonable
+        // minimum.
+        this.targetSize = Math.max(MIN_SPLIT_SIZE, Math.min(totalSize / parallelism, splitSize));
+
+        CloseableIterable<T> tasksToReplay = CloseableIterable.withNoopClose(readAheadTasks);
+        CloseableIterable<T> allTasks;
+        if (!tasksIter.hasNext()) {
+          allTasks = tasksToReplay;
+
+        } else {
+          CloseableIterable<T> remainingTasks =
+              new CloseableIterable<T>() {
+                @Override
+                public CloseableIterator<T> iterator() {
+                  return tasksIter;
+                }
+
+                @Override
+                public void close() throws IOException {
+                  tasksIter.close();
```

Comment:
is this needed? `taskIter` is also closed in the finally block at line 439

Reply:
Yes, this is needed. The close on 439 closes if there is an exception or if the iterable has been exhausted (the other branch of the `tasksIter.hasNext()` check). But if an iterable is returned by this block, the responsibility to close passes to the caller and `shouldClose` is set to false on line 432.